### ISSUE TITLE
RFC: Image Generation Support (Stable Diffusion, Flux, DiT)

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/PyGPUkit#177]
# RFC: Image Generation Model Support (Stable Diffusion, Flux, etc.)

## Summary

PyGPUkitに画像生成モデル（Stable Diffusion, Flux, PixArt等）のサポートを追加するための基盤設計。

### Target Models

| Model | Architecture | Size | VRAM (1024x1024) |
|-------|-------------|------|------------------|
| SD 1.5 | UNet | ~4GB | 6-8GB |
| SDXL | UNet | ~6.9GB | 12-16GB |
| SD3/3.5 | DiT (MMDiT) | 2.5-8B | 8-24GB |
| Flux.1 Dev | DiT | 12B | 16-24GB |
| PixArt-Sigma | DiT | ~10GB | 12-16GB |
| Playground v2.5 | UNet (SDXL) | ~6.9GB | 12-16GB |

---

### Implemented Kernels

| Kernel | Status | Notes |
|--------|--------|-------|
| MatMul (GEMM/GEMV) | ✅ | TF32 TensorCore, cuBLASLt |
| LayerNorm | ✅ | FP32/FP16/BF16 |
| RMSNorm | ✅ | FP32/FP16/BF16 |
| SDPA (causal) | ✅ | CUDA Graph support |
| RoPE (1D) | ✅ | FP32 table variant |
| GELU / SiLU | ✅ | |
| Embedding | ✅ | |
| SafeTensors | ✅ | Zero-copy loading |

### Missing Kernels (for Image Generation)

| Kernel | Priority | Usage |
|--------|----------|-------|
| **GroupNorm** | CRITICAL | VAE, UNet ResNet blocks |
| **2D Convolution** | CRITICAL | UNet backbone, VAE |
| **Cross-Attention** | HIGH | Text-to-image conditioning |
| **AdaLN / AdaLN-Zero** | HIGH | DiT timestep conditioning |
| **Timestep Embedding** | MEDIUM | Sinusoidal encoding |
| **2D RoPE** | MEDIUM | Flux, modern DiT |
| **Transpose Conv** | MEDIUM | VAE decoder upsampling |
| FlashAttention | HIGH | Long sequence efficiency |

---

### 1. Diffusion Model Components

```
Text Encoder (CLIP/T5)
       ↓
    [text_embeds]
       ↓
Noise Sampler → UNet/DiT → Denoised Latent → VAE Decoder → Image
       ↑           ↑
  timestep    cross-attn
```

### 2. UNet Architecture (SD 1.5, SDXL)

```python
@dataclass(frozen=True)
class UNetSpec:
    """Specification for UNet-based diffusion models."""
    name: str

    # Encoder/Decoder block patterns
    down_blocks: list[str]  # "CrossAttnDownBlock2D", "DownBlock2D"
    up_blocks: list[str]    # "CrossAttnUpBlock2D", "UpBlock2D"
    mid_block: str          # "UNetMidBlock2DCrossAttn"

    # Dimensions
    in_channels: int = 4    # Latent channels
    out_channels: int = 4
    block_out_channels: tuple[int, ...] = (320, 640, 1280, 1280)

    # Attention
    cross_attention_dim: int = 768  # Text encoder dim
    attention_head_dim: int = 8

    # Normalization
    norm_num_groups: int = 32  # GroupNorm groups
    norm_eps: float = 1e-5

SDXL_UNET_SPEC = UNetSpec(
    name="sdxl_unet",
    down_blocks=["CrossAttnDownBlock2D"] * 3 + ["DownBlock2D"],
    up_blocks=["UpBlock2D"] + ["CrossAttnUpBlock2D"] * 3,
    mid_block="UNetMidBlock2DCrossAttn",
    block_out_channels=(320, 640, 1280, 1280),
    cross_attention_dim=2048,  # Dual CLIP
)
```

### 3. DiT Architecture (SD3, Flux)

```python
@dataclass(frozen=True)
class DiTSpec:
    """Specification for Diffusion Transformer models."""
    name: str

    # Core dimensions
    hidden_size: int
    num_layers: int
    num_heads: int

    # Conditioning
    conditioning_type: Literal["adaln", "adaln_zero", "cross_attn"]
    text_encoder_dim: int

    # Position encoding
    pos_embed_type: Literal["sinusoidal", "rope_2d", "learned"]
    patch_size: int = 2  # Latent patch size

    # MMDiT specific
    is_mmdit: bool = False  # Multi-modal DiT (SD3)

SD3_DIT_SPEC = DiTSpec(
    name="sd3_mmdit",
    hidden_size=1536,
    num_layers=24,
    num_heads=24,
    conditioning_type="adaln_zero",
    text_encoder_dim=4096,  # T5-XXL
    pos_embed_type="rope_2d",
    is_mmdit=True,
)

FLUX_DIT_SPEC = DiTSpec(
    name="flux_dit",
    hidden_size=3072,
    num_layers=19,  # Double blocks
    num_heads=24,
    conditioning_type="adaln",
    text_encoder_dim=4096,  # T5-XXL
    pos_embed_type="rope_2d",
)
```

### 4. VAE Architecture

```python
@dataclass(frozen=True)
class VAESpec:
    """Specification for VAE encoder/decoder."""
    name: str

    # Dimensions
    in_channels: int = 3
    out_channels: int = 3
    latent_channels: int = 4

    # Scaling
    scaling_factor: float = 0.18215  # SD 1.5
    # scaling_factor: float = 0.13025  # SDXL

    # Architecture
    block_out_channels: tuple[int, ...] = (128, 256, 512, 512)
    layers_per_block: int = 2
    norm_num_groups: int = 32

SDXL_VAE_SPEC = VAESpec(
    name="sdxl_vae",
    latent_channels=4,
    scaling_factor=0.13025,
    block_out_channels=(128, 256, 512, 512),
)
```

### 5. Text Encoder Specs

```python
@dataclass(frozen=True)
class TextEncoderSpec:
    """Specification for text encoders."""
    name: str
    encoder_type: Literal["clip", "t5", "bert"]
    hidden_size: int
    max_length: int

    # Weight patterns
    weight_patterns: dict[str, str]

CLIP_L_SPEC = TextEncoderSpec(
    name="clip_l",
    encoder_type="clip",
    hidden_size=768,
    max_length=77,
    weight_patterns={
        "text_model": "text_model.encoder.layers.{layer}",
        "final_layer_norm": "text_model.final_layer_norm",
    },
)

T5_XXL_SPEC = TextEncoderSpec(
    name="t5_xxl",
    encoder_type="t5",
    hidden_size=4096,
    max_length=512,
    weight_patterns={
        "encoder": "encoder.block.{layer}",
    },
)
```

---

#### 1.1 GroupNorm

```cpp
// native/ops/groupnorm.cu

// NCHW layout (simple, initial implementation)
template<typename T>
__global__ void group_norm_nchw(
    const T* input,      // [N, C, H, W]
    const T* gamma,      // [C]
    const T* beta,       // [C]
    T* output,           // [N, C, H, W]
    int N, int C, int H, int W,
    int num_groups,
    float eps
);

// NHWC layout (optimized for TensorCore)
template<typename T>
__global__ void group_norm_nhwc(
    const T* input,      // [N, H, W, C]
    const T* gamma,      // [C]
    const T* beta,       // [C]
    T* output,           // [N, H, W, C]
    int N, int H, int W, int C,
    int num_groups,
    float eps
);

// Fused GroupNorm + SiLU (20-40% memory bandwidth reduction)
template<typename T>
__global__ void group_norm_silu_fused(
    const T* input,
    const T* gamma,
    const T* beta,
    T* output,
    int N, int C, int H, int W,
    int num_groups,
    float eps
);
```

**Optimization Strategy:**
- Start with NCHW (simpler validation)
- Optimize to NHWC for TensorCore efficiency (35% speedup)
- Fuse with SiLU/GELU for VAE decoder

#### 1.2 2D Convolution

```cpp
// native/ops/conv2d.cu

// Basic 2D convolution (use CUTLASS templates)
template<typename T>
void conv2d_forward(
    const T* input,      // [N, C_in, H, W]
    const T* weight,     // [C_out, C_in, K_h, K_w]
    const T* bias,       // [C_out] or nullptr
    T* output,           // [N, C_out, H_out, W_out]
    int N, int C_in, int H, int W,
    int C_out, int K_h, int K_w,
    int stride_h, int stride_w,
    int pad_h, int pad_w
);

// Transpose convolution (for upsampling)
template<typename T>
void conv2d_transpose(
    const T* input,
    const T* weight,
    const T* bias,
    T* output,
    int N, int C_in, int H, int W,
    int C_out, int K_h, int K_w,
    int stride_h, int stride_w,
    int pad_h, int pad_w,
    int output_pad_h, int output_pad_w
);
```

**Implementation Options:**
1. **cuDNN**: Fastest, but adds dependency
2. **CUTLASS conv2d**: Good performance, self-contained
3. **Custom implicit GEMM**: Full control, more work

**Recommendation:** Start with CUTLASS conv2d templates, optimize later.

#### 1.3 Cross-Attention

```cpp
// native/ops/cross_attention.cu

// Cross-attention (Q from image, K/V from text)
template<typename T>
__global__ void cross_attention_forward(
    const T* query,      // [B, H, N_q, D]  (image tokens)
    const T* key,        // [B, H, N_kv, D] (text tokens)
    const T* value,      // [B, H, N_kv, D]
    T* output,           // [B, H, N_q, D]
    int batch_size,
    int num_heads,
    int seq_len_q,
    int seq_len_kv,
    int head_dim,
    float scale
);
```

**Difference from existing SDPA:**
- No causal mask (bidirectional)
- Different sequence lengths for Q vs K/V
- Q = image features, K/V = text embeddings

#### 2.1 AdaLN / AdaLN-Zero

```cpp
// native/ops/adaln.cu

// AdaLN: Adaptive Layer Normalization
// y = (1 + scale) * LayerNorm(x) + shift
template<typename T>
__global__ void adaln_forward(
    const T* x,          // [B, N, D]
    const T* scale,      // [B, D] from timestep MLP
    const T* shift,      // [B, D] from timestep MLP
    T* output,           // [B, N, D]
    int B, int N, int D,
    float eps
);

// AdaLN-Zero: with gate for residual
// y = x + gate * attn(adaln(x))
template<typename T>
__global__ void adaln_zero_forward(
    const T* x,
    const T* scale,
    const T* shift,
    const T* gate,       // [B, D] gating factor
    T* output,
    int B, int N, int D,
    float eps
);
```

#### 2.2 Timestep Embedding

```cpp
// native/ops/timestep_embed.cu

// Sinusoidal timestep embedding
template<typename T>
__global__ void sinusoidal_timestep_embedding(
    const float* timesteps,  // [B]
    T* output,               // [B, D]
    int batch_size,
    int embedding_dim,
    float max_period = 10000.0f
);
```

#### 2.3 2D RoPE (Axial)

```cpp
// native/ops/rope2d.cu

// 2D RoPE for image patches
template<typename T>
__global__ void rope_2d_forward(
    T* query,            // [B, H, N, D]
    T* key,              // [B, H, N, D]
    const float* freqs,  // Pre-computed frequencies
    int batch_size,
    int num_heads,
    int seq_len,
    int head_dim,
    int height,          // Patch grid height
    int width            // Patch grid width
);
```

---

## Module Structure

```
src/pygpukit/
├── diffusion/                    # NEW
│   ├── __init__.py
│   ├── config.py                 # UNetSpec, DiTSpec, VAESpec
│   ├── scheduler/
│   │   ├── __init__.py
│   │   ├── ddpm.py              # DDPM scheduler
│   │   ├── ddim.py              # DDIM scheduler
│   │   ├── euler.py             # Euler/Euler-A
│   │   └── rectified_flow.py    # SD3, Flux
│   ├── models/
│   │   ├── __init__.py
│   │   ├── unet.py              # UNet2DConditionModel
│   │   ├── dit.py               # DiT, MMDiT
│   │   └── vae.py               # AutoencoderKL
│   ├── text_encoders/
│   │   ├── __init__.py
│   │   ├── clip.py              # CLIP text encoder
│   │   └── t5.py                # T5 encoder
│   ├── loader.py                 # load_diffusion_model()
│   └── pipeline.py               # Text2ImagePipeline
│
├── ops/
│   ├── nn.py                     # MODIFY: Add GroupNorm, cross_attention
│   └── conv.py                   # NEW: 2D convolution ops
│
native/
├── ops/
│   ├── groupnorm.cu              # NEW
│   ├── conv2d.cu                 # NEW
│   ├── cross_attention.cu        # NEW
│   ├── adaln.cu                  # NEW
│   ├── timestep_embed.cu         # NEW
│   └── rope2d.cu                 # NEW
```

---

### High-Level Pipeline

```python
from pygpukit.diffusion import Text2ImagePipeline

# Load model (auto-detects architecture)
pipe = Text2ImagePipeline.from_safetensors(
    "F:/StableDiffusion/sdxl-base-1.0",
    dtype="bfloat16",
)

# Generate image
image = pipe(
    prompt="A photo of a cat",
    negative_prompt="blurry, low quality",
    num_inference_steps=30,
    guidance_scale=7.5,
    width=1024,
    height=1024,
    seed=42,
)

# Save result
image.save("output.png")
```

### Low-Level Components

```python
from pygpukit.diffusion import UNet2D, VAE, CLIPTextEncoder
from pygpukit.diffusion.scheduler import EulerDiscreteScheduler
from pygpukit.core import GPUArray

# Load components separately
text_encoder = CLIPTextEncoder.from_safetensors("clip_l.safetensors")
unet = UNet2D.from_safetensors("unet.safetensors")
vae = VAE.from_safetensors("vae.safetensors")
scheduler = EulerDiscreteScheduler(num_train_timesteps=1000)

# Manual inference loop
text_embeds = text_encoder("A photo of a cat")
latents = GPUArray.randn((1, 4, 128, 128), dtype="bfloat16")

scheduler.set_timesteps(30)
for t in scheduler.timesteps:
    noise_pred = unet(latents, t, text_embeds)
    latents = scheduler.step(noise_pred, t, latents)

image = vae.decode(latents)
```

### Kernel-Level Access

```python
from pygpukit.ops import group_norm, conv2d, cross_attention

# GroupNorm
x = GPUArray.randn((1, 320, 64, 64), dtype="bfloat16")
gamma = GPUArray.ones((320,), dtype="bfloat16")
beta = GPUArray.zeros((320,), dtype="bfloat16")
out = group_norm(x, gamma, beta, num_groups=32, eps=1e-5)

# 2D Convolution
x = GPUArray.randn((1, 320, 64, 64), dtype="bfloat16")
weight = GPUArray.randn((640, 320, 3, 3), dtype="bfloat16")
out = conv2d(x, weight, stride=1, padding=1)

# Cross-Attention
q = GPUArray.randn((1, 8, 4096, 64), dtype="bfloat16")  # Image features
k = GPUArray.randn((1, 8, 77, 64), dtype="bfloat16")    # Text features
v = GPUArray.randn((1, 8, 77, 64), dtype="bfloat16")
out = cross_attention(q, k, v)
```

---

### Phase 1: VAE + Basic UNet (v0.3.0)

**Goal:** Enable SD 1.5 VAE encode/decode

1. GroupNorm kernel (NCHW)
2. 2D Convolution kernel (basic)
3. VAE encoder/decoder
4. Image I/O (PNG/JPEG)

**Deliverable:** `vae.encode(image)` / `vae.decode(latents)`

### Phase 2: Full UNet (v0.3.x)

**Goal:** Enable SD 1.5 / SDXL inference

1. Cross-attention kernel
2. Timestep embedding
3. UNet2DConditionModel
4. CLIP text encoder
5. Basic schedulers (DDIM, Euler)

**Deliverable:** `pipe("prompt")` for SD 1.5/SDXL

### Phase 3: DiT Support (v0.4.x)

**Goal:** Enable SD3, Flux inference

1. AdaLN / AdaLN-Zero kernel
2. 2D RoPE kernel
3. DiT / MMDiT model
4. T5 text encoder
5. Rectified Flow scheduler

**Deliverable:** `pipe("prompt")` for SD3/Flux

### Phase 4: Optimizations (v0.5.x)

**Goal:** Match/exceed vendor performance

1. NHWC GroupNorm + Conv (TensorCore)
2. FlashAttention integration
3. Fused kernels (Conv+GroupNorm+SiLU)
4. CUDA Graph for sampling loop
5. Tiled VAE decode (large images)

---

### Peak VRAM by Model

| Model | Resolution | FP16 | BF16 | FP8 (weights) |
|-------|-----------|------|------|---------------|
| SD 1.5 | 512x512 | 6GB | 6GB | 4GB |
| SDXL | 1024x1024 | 12GB | 12GB | 8GB |
| SD3 Medium | 1024x1024 | 10GB | 10GB | 7GB |
| Flux.1 Schnell | 1024x1024 | 16GB | 16GB | 10GB |
| Flux.1 Dev | 1024x1024 | 24GB | 24GB | 14GB |

### Memory Optimization Strategies

1. **VAE Tiling**: Process large images in tiles
2. **Attention Slicing**: Reduce attention memory
3. **CPU Offload**: Move inactive components to CPU
4. **FP8 Quantization**: 50% weight memory reduction

---

### Kernel Performance (vs PyTorch)

| Kernel | Target | Notes |
|--------|--------|-------|
| GroupNorm | 1.5x faster | NHWC + fused activation |
| Conv2D | 1.0x (parity) | CUTLASS baseline |
| Cross-Attention | 2.0x faster | FlashAttention |
| AdaLN | 1.2x faster | Fused with LayerNorm |

### End-to-End Performance

| Model | Steps | PyTorch (RTX 3090) | PyGPUkit Target |
|-------|-------|-------------------|-----------------|
| SDXL | 30 | ~15s | ~12s (1.25x) |
| SD3 | 28 | ~12s | ~10s (1.2x) |
| Flux Schnell | 4 | ~4s | ~3s (1.3x) |

---

## Open Questions

1. **cuDNN vs CUTLASS**: Use cuDNN for conv (faster) or CUTLASS (no dependency)?
2. **Scheduler in Rust**: Should schedulers be Rust-side for memory management?
3. **LoRA Support**: How to handle LoRA weight merging?
4. **ControlNet**: Support in initial release or later?
5. **Video Diffusion**: Scope creep or roadmap item?

---

## References

- [Stable Diffusion Architecture](https://en.wikipedia.org/wiki/Stable_Diffusion)
- [SDXL Paper](https://arxiv.org/abs/2307.01952)
- [Stable Diffusion 3 Paper](https://stability.ai/news/stable-diffusion-3-research-paper)
- [Flux Architecture](https://arxiv.org/html/2507.09595v1)
- [FlashAttention Paper](https://arxiv.org/abs/2205.14135)
- [Scalable Diffusion Models with Transformers (DiT)](https://arxiv.org/abs/2212.09748)
- [CUTLASS Conv2d](https://github.com/NVIDIA/cutlass/tree/main/examples/conv2d)
- [NHWC GroupNorm CUDA](https://github.com/latentCall145/channels-last-groupnorm)
